### DISCUSSION TITLE
Clone cacheless form preview for templates

### DISCUSSION THREAD

**Discussion by jtinsky** [sheltermanager/asm3#1541] [Ideas]
I can see the recent changes in my template in a popup by selecting that template then clicking the "Preview" button but that doesn't allow me to do proper web development where I can use the dev tools to make changes which I then copy back to the template. The way the form preview works ([development mode](https://github.com/sheltermanager/asm3/issues/960)) is much more useful for making changes to make the templates match the look and feel of a client's site. 

Alternatively, adding a querystring parameter to disable the cache would accomplish the same thing without the need to alter the template admin interface.

It looks like the form development mode was added after the [template preview button](https://github.com/sheltermanager/asm3/issues/603) so ideally the template code could just be updated.

**Comment by bobintetley** [comment]
Sorry, I'm not sure I understand what you're getting at? HTML template preview and form development mode work in exactly the same way, they show you the object you're editing without caching it, so you can leave it open in another tab and reload after you've made changes. 

We don't want people to be able to bust the cache with querystring parameters, because it could easily lead to a DoS since each request requires database calls to serve.

**Comment by jtinsky** [comment]
OK, I see now that I need to check the box, then select which animals to
show on the template. It seems like selecting a filter, in my case a flag,
should automatically select those animals but I’ve checked them and now I’m
seeing the template preview. What I’m not seeing at my template preview
<https://us02d.sheltermanager.com/htmltemplates_preview?template=foreverfosters,&animals=849,857,680,866,937,747,1033,982,919,891,1051,787,738,947,993,684,835,1034,703,955,764,1006,897,1035,911,936,855,956,1040&order=name_asc>
are
all of the animals I’ve checked nor am I seeing the ordering parameter
being applied. Is there a limit on the number of animals that can be shown
in the preview? It doesn’t matter that much to me for dev purposes but I
need to know so I can tell the client when I demo it later. I don’t see any
of this in the documentation so please point me to where this is documented
so I’m not asking a bunch of questions that have already been answered.

**Comment by bobintetley** [comment]
htmltemplates_preview is part of the UI - it can only be accessed when logged in, and you can't pass your own querystring parameters to it. The animals are output in the order that they appeared in the animals list. It's a development time tool and not intended for anything else. There's no limit on the number of animals that can be shown, with the $$BODY$$ section repeated for each one, however if your HTML is malformed for any reason, your browser might not render them in the output and you'll see errors at the console.

**Reply by jtinsky** [reply]
I've checked 29 animals and the preview is only returning 10. This is the same in the unaltered default "littlebox" template. Below is a screenshot of the source of the littlebox template where I'm searching for a string in the body template that shows only 10 animals returned. I'm seeing the same thing in the dev tools BTW.

![image](https://github.com/user-attachments/assets/f07c7ad6-5d30-4ca5-a276-5ce164627f16)

Again, if I should be running this through support please let me know and I'll move the conversation there.

**Reply by bobintetley** [reply]
Apologies, I got this wrong. Looking at the code, the template preview limits itself to the first 10 animals. I guess when we originally added this feature, we thought that no-one would need to be previewing with more than that since the objective was just to test the template looks ok with some animals. It's not that it wouldn't be an easy limit for us to remove, but is there a particular reason that you need to preview a template with more than 10 animals?

**Comment by jtinsky** [comment]
Mainly just so that when I show it to the “client” they don’t think it’s
broken. In my experience people who aren’t doing the development frequently
can’t translate “this isn’t how it will look when it’s live” and fill in
the gaps in the visual or user experience. The number of them will throw
them off as much as the lack of sorting too. I get that it’s work with zero
upside for you so if it’s a big lift I’ll wait until the live version cache
has expired and show them the final version there.

**Reply by bobintetley** [reply]
Ah ok, I understand. The server side cache time is only 30 minutes since last request so it shouldn't keep you waiting too long.
